File: backends/ruby-rails/app/domain/staff/repository.rb
# frozen_string_literal: true
#
# スタッフリポジトリインターフェースを定義するドメインオブジェクトモジュール。
#
# @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>

module Domain
  module Staff
    # スタッフリポジトリのインターフェースです。
    module Repository
      # @param cond [Domain::Staff::Condition] 検索条件
      # @return [Integer] 総件数
      def count_by_condition(cond)                  = raise NotImplementedError

      # @param cond [Domain::Staff::Condition] 検索条件
      # @return [Array<Domain::Staff::Entity>] スタッフエンティティの配列
      def find_by_condition(cond)                   = raise NotImplementedError

      # @param id [Integer] スタッフ ID
      # @return [Domain::Staff::Entity] スタッフエンティティ
      def find_by_id(id)                            = raise NotImplementedError

      # @param provider [Integer] プロバイダー種別
      # @param provider_id [String] プロバイダー ID
      # @return [Domain::Staff::Entity, nil] スタッフエンティティ
      def find_by_provider(provider, provider_id)   = raise NotImplementedError

      # @return [Array<Domain::Staff::Entity>] アクティブなスタッフエンティティの配列
      def find_all_active                            = raise NotImplementedError

      # @param entity [Domain::Staff::Entity] 保存するエンティティ
      # @return [Domain::Staff::Entity] 保存後のエンティティ
      def save(entity)                              = raise NotImplementedError

      # @param id [Integer] スタッフ ID
      # @param role [Integer] 新しいロール
      # @param updated_by [Integer] 更新者 ID
      # @return [void]
      def update_role(id, role, updated_by)         = raise NotImplementedError

      # @param id [Integer] スタッフ ID
      # @param deleted_by [Integer] 削除者 ID
      # @param version [Integer] 削除対象の期待バージョン
      # @return [Boolean] 削除成否
      def soft_delete(id, deleted_by, version)      = raise NotImplementedError

      # @param id [Integer] スタッフ ID
      # @return [void]
      def restore(id)                               = raise NotImplementedError
    end
  end
end

File: backends/kotlin-ktor/src/main/kotlin/com/authorization/domain/staff/Repository.kt
/*
 * スタッフドメインリポジトリインターフェースモジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
package com.authorization.domain.staff

/**
 * スタッフリポジトリのインターフェースです。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
interface Repository {

    /**
     * 検索条件に一致するスタッフ一覧を取得します。
     *
     * @param cond 検索条件
     * @return スタッフ一覧
     */
    suspend fun findByCondition(cond: Condition): List<Staff>

    /**
     * 指定した ID のスタッフを取得します。
     *
     * @param id スタッフ ID
     * @return スタッフ、または null
     */
    suspend fun findById(id: Long): Staff?

    /**
     * プロバイダー情報に一致するスタッフを取得します。
     *
     * @param provider プロバイダー種別
     * @param providerId プロバイダー ID
     * @return スタッフ、または null
     */
    suspend fun findByProvider(provider: Int, providerId: String): Staff?

    /**
     * アクティブなスタッフ全員を取得します。
     *
     * @return スタッフ一覧
     */
    suspend fun findAllActive(): List<Staff>

    /**
     * スタッフを保存します（新規登録・更新）。
     *
     * @param s スタッフエンティティ
     * @return 保存後のスタッフエンティティ
     */
    suspend fun save(s: Staff): Staff

    /**
     * スタッフのロールを更新します。
     *
     * @param id スタッフ ID
     * @param role 新しいロール
     * @param updatedBy 更新者スタッフ ID
     * @return 更新成功なら true
     */
    suspend fun updateRole(id: Long, role: Int, updatedBy: Long): Boolean

    /**
     * スタッフを論理削除します。
     *
     * @param id スタッフ ID
     * @param deletedBy 削除者スタッフ ID
     * @return 削除成功なら true
     */
    suspend fun softDelete(id: Long, deletedBy: Long): Boolean

    /**
     * 論理削除されたスタッフを復元します。
     *
     * @param id スタッフ ID
     * @return 復元成功なら true
     */
    suspend fun restore(id: Long): Boolean
}
